### PR TITLE
feat(eks): support for Kubernetes version 1.23

### PR DESCRIPTION
**Pull request by kaizencc**
Add support for Kubernetes Version 1.23. In order to use this version, customers must pass in a `KubectlLayer` object from `@aws-cdk/lambda-layer-kubectl-v23` to the `kubectlLayer` construct prop of `Cluster`.

----

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks/package.json [hunk 1]
```diff
@@ -80,7 +80,8 @@
   },
   "license": "Apache-2.0",
   "devDependencies": {
-    "@aws-cdk/lambda-layer-kubectl-v22": "2.0.0",
+    "@aws-cdk/lambda-layer-kubectl-v22": "^2.0.0",
+    "@aws-cdk/lambda-layer-kubectl-v23": "^2.0.0",
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-eks/package.json:84]: Any reason why we are locking to 2.0.0? Or does adding `^` make sense.

**Comment by TheRealAmazonKendra** [comment]
PR Linter has words for you.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 2]
```diff
@@ -1372,8 +1381,9 @@ export class Cluster extends ClusterBase {
     this.prune = props.prune ?? true;
     this.vpc = props.vpc || new ec2.Vpc(this, 'DefaultVpc');
 
-    if (props.version === KubernetesVersion.V1_22 && !props.kubectlLayer) {
-      Annotations.of(this).addWarning(`You created a cluster with Kubernetes Version ${props.version} without specifying the kubectlLayer property. This may cause failures as the kubectl version provided with aws-cdk-lib is 1.20, which is only guaranteed to be compatible with Kubernetes versions 1.19-1.21. Please provide a kubectlLayer from @aws-cdk/lambda-layer-kubectl-v22.`);
+    const kubectlVersion = Number(props.version.version) * 100;
+    if (kubectlVersion >= 122 && !props.kubectlLayer) {
+      Annotations.of(this).addWarning(`You created a cluster with Kubernetes Version ${props.version} without specifying the kubectlLayer property. This may cause failures as the kubectl version provided with aws-cdk-lib is 1.20, which is only guaranteed to be compatible with Kubernetes versions 1.19-1.21. Please provide a kubectlLayer from @aws-cdk/lambda-layer-kubectl-v${kubectlVersion % 100}.`);
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:1386]: Can you add a unit test for this annotation? I should have added one in the previous PR. That's a miss on my part.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 3]
```diff
@@ -2300,7 +2310,7 @@ export enum CoreDnsComputeType {
   /**
    * Deploy CoreDNS on Fargate-managed instances.
    */
-  FARGATE = 'fargate'
+  FARGATE = 'fargate',
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:2313]: This is an unrelated change, I think revert it. While it is nice to fix the styling, it will mess up the git blame. When new entries are actually added to this enum, someone can do it then.

Same comment for the other 2

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 4]
```diff
@@ -853,6 +853,15 @@ export class KubernetesVersion {
    */
   public static readonly V1_22 = KubernetesVersion.of('1.22');
 
+  /**
+   * Kubernetes version 1.23
+   *
+   * When creating a `Cluster` with this version, you need to also specify the
+   * `kubectlLayer` property with a `KubectlLayer` from
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:860]:
```suggestion
   * `kubectlLayer` property with a `KubectlV23Layer` from
```

**Review by madeline-k — CHANGES_REQUESTED** [review]

**Comment by madeline-k** [comment]
Depends on https://github.com/aws/aws-cdk/pull/22658

**Comment by kaizencc** [comment]
readme exempt; this is a feature for a new version and the method to include new kubectl versions is already included in the readme.

**Comment by robertd** [comment]
@pahud We need MVP awards/recognition for each release. I nominate @madeline-k & @kaizencc for the upcoming v2.48.0 release 😃  🎉 🚀

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 5]
```diff
@@ -1372,8 +1381,9 @@ export class Cluster extends ClusterBase {
     this.prune = props.prune ?? true;
     this.vpc = props.vpc || new ec2.Vpc(this, 'DefaultVpc');
 
-    if (props.version === KubernetesVersion.V1_22 && !props.kubectlLayer) {
-      Annotations.of(this).addWarning(`You created a cluster with Kubernetes Version ${props.version} without specifying the kubectlLayer property. This may cause failures as the kubectl version provided with aws-cdk-lib is 1.20, which is only guaranteed to be compatible with Kubernetes versions 1.19-1.21. Please provide a kubectlLayer from @aws-cdk/lambda-layer-kubectl-v22.`);
+    const kubectlVersion = Number(props.version.version) * 100;
+    if (kubectlVersion >= 122 && !props.kubectlLayer) {
+      Annotations.of(this).addWarning(`You created a cluster with Kubernetes Version ${props.version.version} without specifying the kubectlLayer property. This may cause failures as the kubectl version provided with aws-cdk-lib is 1.20, which is only guaranteed to be compatible with Kubernetes versions 1.19-1.21. Please provide a kubectlLayer from @aws-cdk/lambda-layer-kubectl-v${kubectlVersion % 100}.`);
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:1386]: Is it that it may cause errors or that it will? If we know it will, I think this should be an error.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1386]: If not, no issues here.

File: packages/@aws-cdk/aws-eks/test/alb-controller.integ.snapshot/manifest.json [hunk 6]
```diff
@@ -168,7 +168,10 @@
         "/aws-cdk-eks-cluster-alb-controller-test/KubectlLayer/Resource": [
           {
             "type": "aws:cdk:logicalId",
-            "data": "KubectlLayer600207B5"
+            "data": "KubectlLayer600207B5",
+            "trace": [
+              "!!DESTRUCTIVE_CHANGES: WILL_REPLACE"
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-eks/test/alb-controller.integ.snapshot/manifest.json:173]: Is this a concern?

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 7]
```diff
@@ -1372,8 +1381,9 @@ export class Cluster extends ClusterBase {
     this.prune = props.prune ?? true;
     this.vpc = props.vpc || new ec2.Vpc(this, 'DefaultVpc');
 
-    if (props.version === KubernetesVersion.V1_22 && !props.kubectlLayer) {
-      Annotations.of(this).addWarning(`You created a cluster with Kubernetes Version ${props.version} without specifying the kubectlLayer property. This may cause failures as the kubectl version provided with aws-cdk-lib is 1.20, which is only guaranteed to be compatible with Kubernetes versions 1.19-1.21. Please provide a kubectlLayer from @aws-cdk/lambda-layer-kubectl-v22.`);
+    const kubectlVersion = Number(props.version.version) * 100;
```
Comment (MrArnoldPalmer) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:1384]: I feel we should just use `semver` here to compare versions, no need to parse to a number and *100 stuff.

**Review by MrArnoldPalmer — CHANGES_REQUESTED** [review]
just a small suggestion

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 5 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1386]: It is not for certain. In some cases you can use a version of kubectl that is not guaranteed to be compatible with the version of the Kubernetes engine you are using. It just depends on the APIs being used.

File: packages/@aws-cdk/aws-eks/test/alb-controller.integ.snapshot/manifest.json (hunk 6 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-eks/test/alb-controller.integ.snapshot/manifest.json:173]: No, a lambda layer is a stateless resource, and we are replacing it with an upgraded version of kubectl. So, the replacement is expected here.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 7 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1384]: Okay fair enough, I'm fine with either approach. We can update this.

Comment (MrArnoldPalmer) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1384]: it's already a dependency of aws-cdk-lib so no cost really. Otherwise we will have to parse each individual segment major + minor from both and compare both but since semver already does this that just feels nicer.

Comment (madeline-k) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts]: Updated 👍

**Review by MrArnoldPalmer — APPROVED** [review]

**Review by madeline-k — APPROVED** [review]
